Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -385,314 +372,62 @@ final class ConnectionPool {
-
-                } else {
-                    connection.isLeased = false
-                    self.leased -= 1
-                    if connection.isActiveEstimation, !connection.isClosing {
-                        self.availableConnections.append(connection)
-                    }
-
-                    if self.providerMustClose() {
-                        self.removeFromPool()
-                    }
-
-                    return .none
-                }
-            }
-
-            fileprivate mutating func removeClosedConnection(_ connection: Connection) -> ClosedConnectionRemoveAction {
-                if connection.isLeased {
-                    if let firstWaiter = self.waiters.popFirst() {
-                        let (el, _) = self.resolvePreference(firstWaiter.preference)
-                        return .makeConnectionAndComplete(el, firstWaiter.promise)
-                    }
-                } else {
-                    self.availableConnections.swapWithFirstAndRemove(where: { $0 === connection })
-                }
-
-                if self.providerMustClose() {
-                    self.removeFromPool()
-                }
-
-                return .none
-            }
-
-            fileprivate mutating func failedConnectionAction() -> ClosedConnectionRemoveAction {
-                if let firstWaiter = self.waiters.popFirst() {
-                    let (el, _) = self.resolvePreference(firstWaiter.preference)
-                    return .makeConnectionAndComplete(el, firstWaiter.promise)
-                } else {
-                    self.leased -= 1
-                    if self.providerMustClose() {
-                        self.removeFromPool()
-                    }
-                    return .none
-                }
-            }
-
-            private func providerMustClose() -> Bool {
-                return self.pending == 0 && self.activity != .closed && self.leased == 0 && self.availableConnections.isEmpty && self.waiters.isEmpty
-            }
-
-            /// - Warning: This should always be called from a critical section protected by `.connectionProvidersLock`
-            fileprivate mutating func removeFromPool() {
-                assert(self.parentPool.connectionProviders[self.key] != nil)
-                self.parentPool.connectionProviders[self.key] = nil
-                assert(self.activity != .closed)
-                self.activity = .closed
-            }
-
-            private func resolvePreference(_ preference: HTTPClient.EventLoopPreference) -> (EventLoop, Bool) {
-                switch preference.preference {
-                case .indifferent:
-                    return (self.defaultEventLoop, false)
-                case .delegate(let el):
-                    return (el, false)
-                case .delegateAndChannel(let el), .testOnly_exact(let el, _):
-                    return (el, true)
-                }
-            }
-
-            fileprivate enum ConnectionGetAction {
```

Comment:
These "actions" seem like the right tool to return so that outside of the lock we can then perform the actions.

Reply:
you are right, they are indeed right tool, but I think that State struct maybe not that useful (give a few more hours :D)

Reply:
Yep, they are indeed